File: cpp_merge/classified/packages/plugin-sam/src/actions/sayAloud.cpp
#include "sayAloud.hpp"

string extractTextToSpeak(string messageText)
{
    shared text = messageText->toLowerCase()->trim();
    auto quotedPatterns = array<std::shared_ptr<RegExp>>{ (new RegExp(std:("say ["']([^"']+)["'"))), (new RegExp(std:("speak ["']([^"']+)["'"))), (new RegExp(std:("read ["']([^"']+)["'"))), (new RegExp(std:("announce ["']([^"']+)["'"))), (new RegExp(std:("["']([^"']+)["'"))) };
    for (auto& pattern : quotedPatterns)
    {
        auto match = text->match(pattern);
        if (match) {
            return (*const_(match))[1];
        }
    }
    auto afterKeywordPatterns = array<std::shared_ptr<RegExp>>{ (new RegExp(std:("(?:say|speak|read)\s+(?:aloud\s+)?(?:this\s+)?:?\s*(.+)"))), (new RegExp(std:("(?:can you|please)\s+(?:say|speak|read)\s+(?:aloud\s+)?(.+)"))), (new RegExp(std:("(?:i want to hear|let me hear)\s+(.+)"))), (new RegExp(std:("(?:read this|say this|speak this)\s*:?\s*(.+)"))) };
    for (auto& pattern : afterKeywordPatterns)
    {
        auto match = text->match(pattern);
        if (match) {
            auto extractedText = (*const_(match))[1]->trim();
            extractedText = extractedText->replace((new RegExp(std:("\s+out loud"))), string_empty)->replace((new RegExp(std:("\s+aloud"))), string_empty)->replace((new RegExp(std:("\s+please"))), string_empty)->trim();
            return extractedText;
        }
    }
    auto speechKeywords = array<string>{ std:("say aloud"), std:("speak"), std:("use your voice"), std:("talk to me") };
    if (speechKeywords->some([=](auto keyword) mutable
    {
        return text->includes(keyword);
    }
    )) {
        return std:("Hello! I am speaking using my SAM voice synthesizer.");
    }
    return string_empty;
};


std::shared_ptr<SamTTSOptions> extractVoiceOptions(string messageText)
{
    auto text = messageText->toLowerCase();
    auto options = object{};
    if (OR((OR((text->includes(std:("higher voice"))), (text->includes(std:("high pitch"))))), (text->includes(std:("squeaky"))))) {
        options->pitch = 100;
    } else if (OR((OR((text->includes(std:("lower voice"))), (text->includes(std:("low pitch"))))), (text->includes(std:("deep voice"))))) {
        options->pitch = 30;
    }
    if (OR((OR((text->includes(std:("faster"))), (text->includes(std:("quickly"))))), (text->includes(std:("speed up"))))) {
        options->speed = 120;
    } else if (OR((OR((text->includes(std:("slower"))), (text->includes(std:("slowly"))))), (text->includes(std:("slow down"))))) {
        options->speed = 40;
    }
    if (OR((text->includes(std:("robotic"))), (text->includes(std:("robot voice"))))) {
        options->throat = 200;
        options->mouth = 50;
    } else if (OR((text->includes(std:("smooth"))), (text->includes(std:("natural"))))) {
        options->throat = 100;
        options->mouth = 150;
    }
    return options;
};


std::shared_ptr<Action> sayAloudAction = object{
    object::pair{std:("name"), std:("SAY_ALOUD")}, 
    object::pair{std:("description"), std:("Make the agent speak text aloud using SAM retro speech synthesizer")}, 
    object::pair{std:("examples"), array<array<any>>{ array<object>{ object{
        object::pair{std:("name"), std:("{{user1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Can you say hello out loud?")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{agent}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I'll say hello using my SAM voice.")}, 
            object::pair{std:("action"), std:("SAY_ALOUD")}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("{{user1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Please read this message aloud: Welcome to ElizaOS")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{agent}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I'll read that message aloud for you now.")}, 
            object::pair{std:("action"), std:("SAY_ALOUD")}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("{{user1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Speak in a higher voice")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{agent}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I'll adjust my voice settings and speak in a higher pitch.")}, 
            object::pair{std:("action"), std:("SAY_ALOUD")}
        }}
    } } }}, 
    object::pair{std:("validate"), [=](auto runtime, auto message, auto _state = undefined) mutable
    {
        shared text = message->content->text->toLowerCase();
        auto speechTriggers = array<string>{ std:("say aloud"), std:("speak"), std:("read aloud"), std:("say out loud"), std:("voice"), std:("speak this"), std:("say this"), std:("read this"), std:("announce"), std:("proclaim"), std:("tell everyone"), std:("speak up"), std:("use your voice"), std:("talk to me"), std:("higher voice"), std:("lower voice"), std:("change voice"), std:("robotic voice"), std:("retro voice") };
        auto hasSpeechTrigger = speechTriggers->some([=](auto trigger) mutable
        {
            return text->includes(trigger);
        }
        );
        auto hasVocalizationIntent = OR((OR((OR((OR((OR((text->includes(std:("can you say"))), (text->includes(std:("please say"))))), (text->includes(std:("i want to hear"))))), (text->includes(std:("let me hear"))))), (text->match((new RegExp(std:("say ["'].*["'")))) != nullptr))), (text->match((new RegExp(std:("speak ["'].*["'")))) != nullptr));
        return OR((hasSpeechTrigger), (hasVocalizationIntent));
    }
    }, 
    object::pair{std:("handler"), [=](auto runtime, auto message, auto _state = undefined, auto _options = undefined, auto callback = undefined) mutable
    {
        logger->info(std:("[SAY_ALOUD] Processing speech request..."));
        auto samService = as<any>((as<object>(runtime))["getService"](std:("SAM_TTS")));
        if (!samService) {
            logger->warn(std:("[SAY_ALOUD] SAM TTS service not available"));
            callback(object{
                object::pair{std:("text"), std:("Sorry, I cannot speak aloud right now. The text-to-speech service is not available.")}, 
                object::pair{std:("action"), std:("SAY_ALOUD")}, 
                object::pair{std:("error"), std:("SAM TTS service not available")}
            });
            return std::shared_ptr<Promise<void>>();
        }
        auto textToSpeak = extractTextToSpeak(message->content->text);
        auto voiceOptions = extractVoiceOptions(message->content->text);
        logger->info(std:("[SAY_ALOUD] Speaking: "") + textToSpeak + std:("""));
        logger->info(std:("[SAY_ALOUD] Voice options:"), voiceOptions);
        auto audioBuffer = std::async([=]() { samService->speakText(textToSpeak, voiceOptions); });
        logger->info(std:("[SAY_ALOUD] ✅ Speech synthesis completed successfully"));
        callback(object{
            object::pair{std:("text"), std:("I spoke aloud using my SAM voice: "") + textToSpeak + std:(""")}, 
            object::pair{std:("action"), std:("SAY_ALOUD")}, 
            object::pair{std:("audioData"), Array->from(audioBuffer)}
        });
    }
    }
};

void Main(void)
{
}

MAIN

File: classified/packages/plugin-sam/src/actions/sayAloud.ts
import type { Action, IAgentRuntime, Memory, State, HandlerCallback } from '@elizaos/core';
import { logger } from '@elizaos/core';
import { SamTTSService, type SamTTSOptions } from '../services/SamTTSService';

/**
 * SAY_ALOUD Action
 *
 * Makes the agent speak text aloud using the SAM speech synthesizer.
 * The audio is sent through the hardware bridge to the user's speakers.
 */
export const sayAloudAction: Action = {
  name: 'SAY_ALOUD',
  description: 'Make the agent speak text aloud using SAM retro speech synthesizer',

  examples: [
    [
      {
        name: '{{user1}}',
        content: { text: 'Can you say hello out loud?' },
      },
      {
        name: '{{agent}}',
        content: {
          text: "I'll say hello using my SAM voice.",
          action: 'SAY_ALOUD',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Please read this message aloud: Welcome to ElizaOS' },
      },
      {
        name: '{{agent}}',
        content: {
          text: "I'll read that message aloud for you now.",
          action: 'SAY_ALOUD',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Speak in a higher voice' },
      },
      {
        name: '{{agent}}',
        content: {
          text: "I'll adjust my voice settings and speak in a higher pitch.",
          action: 'SAY_ALOUD',
        },
      },
    ],
  ],

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const text = message.content.text.toLowerCase();

    // Trigger words and phrases for speech synthesis
    const speechTriggers = [
      'say aloud',
      'speak',
      'read aloud',
      'say out loud',
      'voice',
      'speak this',
      'say this',
      'read this',
      'announce',
      'proclaim',
      'tell everyone',
      'speak up',
      'use your voice',
      'talk to me',
      'higher voice',
      'lower voice',
      'change voice',
      'robotic voice',
      'retro voice',
    ];

    // Check if message contains speech-related triggers
    const hasSpeechTrigger = speechTriggers.some((trigger) => text.includes(trigger));

    // Also check if the user is asking agent to vocalize something
    const hasVocalizationIntent =
      text.includes('can you say') ||
      text.includes('please say') ||
      text.includes('i want to hear') ||
      text.includes('let me hear') ||
      text.match(/say ["'].*["']/) !== null ||
      text.match(/speak ["'].*["']/) !== null;

    return hasSpeechTrigger || hasVocalizationIntent;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<void> => {
    logger.info('[SAY_ALOUD] Processing speech request...');

    // Get SAM TTS service
    const samService = (runtime as { getService: (serviceType: string) => unknown }).getService(
      'SAM_TTS'
    ) as SamTTSService | null;

    if (!samService) {
      logger.warn('[SAY_ALOUD] SAM TTS service not available');
      callback({
        text: 'Sorry, I cannot speak aloud right now. The text-to-speech service is not available.',
        action: 'SAY_ALOUD',
        error: 'SAM TTS service not available',
      });
      return;
    }

    // Extract text to speak from the message
    const textToSpeak = extractTextToSpeak(message.content.text);

    // Extract voice options from the message
    const voiceOptions = extractVoiceOptions(message.content.text);

    logger.info(`[SAY_ALOUD] Speaking: "${textToSpeak}"`);
    logger.info('[SAY_ALOUD] Voice options:', voiceOptions);

    // Synthesize and play the speech
    const audioBuffer = await samService.speakText(textToSpeak, voiceOptions);

    logger.info('[SAY_ALOUD] ✅ Speech synthesis completed successfully');

    callback({
      text: `I spoke aloud using my SAM voice: "${textToSpeak}"`,
      action: 'SAY_ALOUD',
      audioData: Array.from(audioBuffer), // Convert Uint8Array to regular array for JSON
    });
  },
};

/**
 * Extract text to speak from user message
 */
function extractTextToSpeak(messageText: string): string {
  // Clean the input
  const text = messageText.toLowerCase().trim();

  // Patterns to extract quoted text
  const quotedPatterns = [
    /say ["']([^"']+)["']/,
    /speak ["']([^"']+)["']/,
    /read ["']([^"']+)["']/,
    /announce ["']([^"']+)["']/,
    /["']([^"']+)["']/,
  ];

  // Try to find quoted text first
  for (const pattern of quotedPatterns) {
    const match = text.match(pattern);
    if (match) {
      return match[1];
    }
  }

  // Patterns to extract text after keywords
  const afterKeywordPatterns = [
    /(?:say|speak|read)\s+(?:aloud\s+)?(?:this\s+)?:?\s*(.+)$/,
    /(?:can you|please)\s+(?:say|speak|read)\s+(?:aloud\s+)?(.+)$/,
    /(?:i want to hear|let me hear)\s+(.+)$/,
    /(?:read this|say this|speak this)\s*:?\s*(.+)$/,
  ];

  // Try to find text after keywords
  for (const pattern of afterKeywordPatterns) {
    const match = text.match(pattern);
    if (match) {
      let extractedText = match[1].trim();

      // Clean up common trailing words
      extractedText = extractedText
        .replace(/\s+out loud$/, '')
        .replace(/\s+aloud$/, '')
        .replace(/\s+please$/, '')
        .trim();

      return extractedText;
    }
  }

  // If no specific text found but speech is requested, use a default response
  const speechKeywords = ['say aloud', 'speak', 'use your voice', 'talk to me'];
  if (speechKeywords.some((keyword) => text.includes(keyword))) {
    return 'Hello! I am speaking using my SAM voice synthesizer.';
  }

  return '';
}

/**
 * Extract voice configuration options from user message
 */
function extractVoiceOptions(messageText: string): SamTTSOptions {
  const text = messageText.toLowerCase();
  const options: SamTTSOptions = {};

  // Pitch modifications
  if (text.includes('higher voice') || text.includes('high pitch') || text.includes('squeaky')) {
    options.pitch = 100; // Higher pitch
  } else if (
    text.includes('lower voice') ||
    text.includes('low pitch') ||
    text.includes('deep voice')
  ) {
    options.pitch = 30; // Lower pitch
  }

  // Speed modifications
  if (text.includes('faster') || text.includes('quickly') || text.includes('speed up')) {
    options.speed = 120; // Faster speech
  } else if (text.includes('slower') || text.includes('slowly') || text.includes('slow down')) {
    options.speed = 40; // Slower speech
  }

  // Voice character modifications
  if (text.includes('robotic') || text.includes('robot voice')) {
    options.throat = 200; // More robotic
    options.mouth = 50; // Sharper articulation
  } else if (text.includes('smooth') || text.includes('natural')) {
    options.throat = 100; // Smoother
    options.mouth = 150; // Softer articulation
  }

  return options;
}
